source/extensions/filters/http/transformation/inja_transformer.cc

```diff
@@ -244,6 +253,40 @@ json TransformerInstance::cluster_metadata_callback(
   return "";
 }
 
+json TransformerInstance::base64_encode_callback(const inja::Arguments &args) const {
+  const std::string &input = args.at(0)->get_ref<const std::string &>();
+  return Base64::encode(input.c_str(), input.length());
+}
+
+json TransformerInstance::base64_decode_callback(const inja::Arguments &args) const {
+  const std::string &input = args.at(0)->get_ref<const std::string &>();
+  return Base64::decode(input);
+}
+
+json TransformerInstance::substring_callback(const inja::Arguments &args) const {
+  const std::string &input = args.at(0)->get_ref<const std::string &>();
+  const int64_t start = args.at(1)->get_ref<const int64_t &>();
+  const int64_t substring_len = args.at(2)->get_ref<const int64_t &>();
+
+  // store input length 
+  const int64_t input_len = input.length();
+
+  // if start is negative, substring_len is non-positive, or 
+  // start is greater than the length of the string, return empty string
+  if (start < 0 || substring_len <= 0 || start >= input_len) {
```

Comment:
should we handle substring_len differently if it's zero? e.g. do what the c++ function does and just take until the end? otherwise it seems like the UX is a bit poor for users who want to take substring from start to end without specifying the length

Reply:
https://github.com/solo-io/envoy-gloo/pull/184/commits/91225d5a1a4b1c320c19f6b964f45ce23ffe8256
Updated this here, let me know what you think!

Reply:
what are some typical ways that negative numbers are used for substringing? i know that in python, slicing with negative values uses indices from the end, for example:

```
>>> test_str = '0123456789'
>>> print(test_str[-5:-2]) # start from 5 before end, end at 2 before end
567
```

but in python's case, the indices mean `start` and `end` rather than `start` and `length`. i don't have experience with substringing using `start` and `length`. but it does feel a bit redundant to do the same thing with all negative values, and i wonder if we could do something more flexible with them. any thoughts?

Reply:
The implementation exposed in this PR is pretty similar to the C++ `<string>.substr` function that we call into in the `substring_callback`
https://github.com/solo-io/envoy-gloo/blob/8ddf225f48d8db0fc7f86aa9c62411797967568e/source/extensions/filters/http/transformation/inja_transformer.cc#L294
where the arguments to the substring function are 
1. the nonnegative start position of the substring
2. (optional) the number of characters in the substring

(We also sometimes use the Google Abseil version of this which operates on a `string_view` and has the same style of signature, see here:
https://github.com/solo-io/envoy-gloo/blob/8ddf225f48d8db0fc7f86aa9c62411797967568e/source/common/matcher/solo_matcher.cc#L99

I'm in favor of maintaining this implementation instead of switching to the python style negative substringing for a couple of reasons:
1. Because we don't (at least as far as I know) have a substring utility that we can call into with python style arguments, implementing the python style substringing is going to be more verbose and therefore harder to debug/maintain going forward than just calling into the available C++ utility
2. I don't think the use case that we're implementing this feature for requires a particularly sophisticated set of substringing tools -- I don't have any info beyond what is given in the comments on the issue, but I would imagine that the workflow is generally something like:
  - decode the base64 string, resulting in a constant length auth token
  - remove a certain known number of characters from the start or end of the string

Reply:
Sorry if that's not a super direct answer. Long story short, I prefer keeping the implementation simple/close to the C++ utilities we're using because 
 - I think it will be easier to maintain
 - it doesn't seem like the feature requester requires a particularly sophisticated set of substringing tools

Reply:
Closing this thread after our discussion in the Gloo Slack:
https://solo-io-corp.slack.com/archives/C03MFATU265/p1664386328617359